### ISSUE TITLE
Re-open: GZIP content not inflated

### ISSUE THREAD

**Issue by phleemac** [loopj/android-async-http#932]
I encounter exactly the same problem as described in https://github.com/loopj/android-async-http/issues/641 using version 1.4.8 

This is caused by a bug in <code>com.loopj.android.http.AsyncHttpClient#isInputStreamGZIPCompressed</code>: in the statement <code>int readStatus = inputStream.read(signature);</code> where <code>signature</code> might only be partially filled (see javadoc http://docs.oracle.com/javase/7/docs/api/java/io/PushbackInputStream.html#read(byte[],%20int,%20int)

Should check the returned byte count and re-read to fetch the 2nd byte if necessary

**Comment by phleemac** [comment]
I temporarily workaround the problem by inserting another response interceptor to pre-fetch the first two bytes.

For reference, in Scala:
<pre>
  private lazy val client = {
    val c = new AsyncHttpClient

    // TODO: make sure at least 2 bytes are available in the content input stream to workaround 
    // https://github.com/loopj/android-async-http/issues/932 (android-async-http 1.4.8)

    c.getHttpClient match {
      case httpClient: DefaultHttpClient =>
        httpClient.addResponseInterceptor(new HttpResponseInterceptor {
          override def process(response: HttpResponse, httpContext: HttpContext): Unit = {
            val entity: HttpEntity = response.getEntity
            if (entity == null) {
              return
            } else {
              response.setEntity(new HttpEntityWrapper(entity) {
                override def getContent: InputStream = {
                  val wrappedStream = wrappedEntity.getContent
                  val pushbackStream = new PushbackInputStream(wrappedStream, 2)
                  def preRead(cnt: Int, buf: Seq[Int]): Unit = {
                    lazy val b = pushbackStream.read()
                    if ((cnt > 0) && (b >= 0)) {
                      preRead(cnt - 1, b +: buf)
                    } else {
                      buf.foreach(pushbackStream.unread(_))
                    }
                  }
                  preRead(2, Seq.empty)
                  pushbackStream
                }
              })
            }
          }
        }, 0)
      case _ => // skip
    }
    c
  }
</pre>

**Comment by Queatz** [comment]
+1 Running into this as well with the Google Autocomplete API.

**Comment by smarek** [comment]
Thank you, I'm able to reproduce this issue, and I'll push the fix asap

**Comment by smarek** [comment]
Simple fix is patch like this
```diff
diff --git a/library/src/main/java/com/loopj/android/http/AsyncHttpClient.java b/library/src/main/java/com/loopj/android/http/AsyncHttpClient.java
index b9ee313..b10ea6f 100755
--- a/library/src/main/java/com/loopj/android/http/AsyncHttpClient.java
+++ b/library/src/main/java/com/loopj/android/http/AsyncHttpClient.java
@@ -1608,7 +1608,8 @@ public class AsyncHttpClient {
         public InputStream getContent() throws IOException {
             wrappedStream = wrappedEntity.getContent();
             pushbackStream = new PushbackInputStream(wrappedStream, 2);
-            if (isInputStreamGZIPCompressed(pushbackStream)) {
+            Header enc = wrappedEntity.getContentEncoding();
+            if ((enc != null && "gzip".equalsIgnoreCase(enc.getValue())) || isInputStreamGZIPCompressed(pushbackStream)) {
                 gzippedStream = new GZIPInputStream(pushbackStream);
                 return gzippedStream;
             } else {
```

But this is not final solution, and will be pushed in better form

**Comment by SergeyA** [comment]
You can disable gzip as a temporary fix:

    httpClient = new AsyncHttpClient();
    httpClient.addHeader("Accept-Encoding", "identity"); // disable gzip
    httpClient.get(...)

**Comment by sonhaxatacmoi** [comment]
@SergeyA I tried your way, with values "identity","",... even remove this header, not working at all.

**Comment by phleemac** [comment]
Thanks @smarek.  But your patch is doing redundant check on the contentEncoding.  It's already checked at around line 280:

<pre>
...
                final Header encoding = entity.getContentEncoding();
                if (encoding != null) {
                    for (HeaderElement element : encoding.getElements()) {
                      <b>  if (element.getName().equalsIgnoreCase(ENCODING_GZIP)) {</b>
                            response.setEntity(new InflatingEntity(entity));
                            break;
                        }
                    }
                }
...
</pre>

Thus the patch effectively disable the call to <code>isInputStreamGZIPCompressed</code>.  In that case, why don't you just remove the call to <code>isInputStreamGZIPCompressed</code>?

If we want to play safe and keep the double checking,  <code>isInputStreamGZIPCompressed</code> should be fixed like:

<pre>
    public static boolean isInputStreamGZIPCompressed(final PushbackInputStream inputStream) throws IOException {
        if (inputStream == null)
            return false;

        byte[] signature = new byte[2];
<b>        int count = 0;
        try {
            while (count < 2) {
                int readCount = inputStream.read(signature, count, 2 - count);
                if (readCount < 0) return false;
                count = count + readCount;
            }
        } finally {
            inputStream.unread(signature, 0, count);
        }    </b>
        int streamHeader = ((int) signature[0] & 0xff) | ((signature[1] << 8) & 0xff00);
        return readStatus == 2 && GZIPInputStream.GZIP_MAGIC == streamHeader;
    }
</pre>

**Comment by smarek** [comment]
@SergeyA thanks for adding workaround til the fix is published in stable
@phleemac yes, that's correct, it does assume that stream is gzip compressed without checking the starting two-byte signature, however the check was introduced with #583 which is meaningfull for most usages, so I'll try to implement that `wait for two bytes on input` procedure as correct solution

**Comment by smarek** [comment]
Thank you nicely for provided patch @phleemac, I've included that in commit 4bffe5f283c3421348d7bc5ae307733f1db0bd31, which will be released in 1.4.9 as soon as possible.
